_posts/2022-02-24-torchserve-grpc.md

```diff
@@ -0,0 +1,217 @@
+
+
+## Torchserve benchmark
+
+I have been using [TorchServe](https://pytorch.org/serve/) for a while now and I am quite happy with it. It provides all the flexibility I need and it is quite simple to set up. Model handlers allow you to customize every detail for your specific model without really worrying about other complex things as batching and queing requests. 
+I do not pretend to give an overview of TorchServe or make a comparison of its advantages compared to other inference servers, I will let that for a plausible future post.
+
+The documentation for Torchserve's [gRPC API](https://pytorch.org/serve/grpc_api.html){:target="_blank"}{:rel="noopener noreferrer"} could be improved, as it currently requires users to download the official repository to generate a Python gRPC client stub from the proto files. However, I have attached these files to the repository, so you can easily run the benchmark without having to worry about this step.
+
+The experiment is very similar to the previous one, sending 20 concurrent request and repeating that 10 times to measure the average time. I am going to use one of the pytorch vision model examples, [densenet161](https://pytorch.org/hub/pytorch_vision_densenet/){:target="_blank"}{:rel="noopener noreferrer"}. The model is not important here since we do not really care about inference results. Let's see some results:
+
+
+<div class="table-wrapper" markdown="block">
+
+|      | B64 (0.306 MB) | Binary (0.229) |
+|------|----------------|----------------|
+| REST | 0.884          | 0.628          |
+| gRPC |       X        | 0.645          |
+
+<p>Table 4. Results for small images: 360x640</p>
+
+</div>
+
+<div class="table-wrapper" markdown="block">
+
+|      | B64 (0.306 MB) | Binary (0.229) |
+|------|----------------|----------------|
+| REST | 1.262          | 0.946          |
+| gRPC |        X       | 0.927          |
+
+<p>Table 5. Results for medium images: 720x1280</p>
+
+</div>
+
+<div class="table-wrapper" markdown="block">
+
+|      | B64 (0.306 MB) | Binary (0.229) |
+|------|----------------|----------------|
+| REST | 2.188          | 1.384          |
+| gRPC |         X      | 1.422          |
+
+<p>Table 6. Results for large images: 1080x1920</p>
+
+</div>
+
+Translating the insights gained from benchmarking with the base servers can be challenging. The tables indicate that Base64 encoding should be avoided and that there are no significant performance differences between using gRPC and REST.
+
+Two factors contribute to the similar performance results for gRPC and REST. Firstly, the model's inference time is considerably longer than the networking time, making it difficult to discern the small gains obtained by changing the transmission protocol. For example, sending 20 large images concurrently in the simple base case (Table 3) took roughly 0.19s, whereas we are now spending approximately 1.4 seconds, highlighting the significant impact of model inference time on the comparison.
```

Comment:
whereas we are now spending approximately 1.4 seconds (Table 6)

Yo mencionaria la tabla para que se vea qué dos cosas hay que comparar